src/main/java/org/opensearch/ad/feature/CompositeRetriever.java

```diff
@@ -0,0 +1,346 @@
+ *
+ */
+public class CompositeRetriever extends AbstractRetriever {
+    private static final String AGG_NAME_COMP = "comp_agg";
+    private static final Logger LOG = LogManager.getLogger(CompositeRetriever.class);
+
+    private final long dataStartEpoch;
+    private final long dataEndEpoch;
+    private final AnomalyDetector anomalyDetector;
+    private final NamedXContentRegistry xContent;
+    private final Client client;
+    private int totalResults;
+    private int maxEntities;
+    private final int pageSize;
+    private long expirationEpochMs;
+    private Clock clock;
+
+    public CompositeRetriever(
+        long dataStartEpoch,
+        long dataEndEpoch,
+        AnomalyDetector anomalyDetector,
+        NamedXContentRegistry xContent,
+        Client client,
+        long expirationEpochMs,
+        Clock clock,
+        Settings settings,
+        int maxEntitiesPerInterval,
+        int pageSize
+    ) {
+        this.dataStartEpoch = dataStartEpoch;
+        this.dataEndEpoch = dataEndEpoch;
+        this.anomalyDetector = anomalyDetector;
+        this.xContent = xContent;
+        this.client = client;
+        this.totalResults = 0;
+        this.maxEntities = maxEntitiesPerInterval;
+        this.pageSize = pageSize;
+        this.expirationEpochMs = expirationEpochMs;
+        this.clock = clock;
+    }
+
+    // a constructor that provide default value of clock
+    public CompositeRetriever(
+        long dataStartEpoch,
+        long dataEndEpoch,
+        AnomalyDetector anomalyDetector,
+        NamedXContentRegistry xContent,
+        Client client,
+        long expirationEpochMs,
+        Settings settings,
+        int maxEntitiesPerInterval,
+        int pageSize
+    ) {
+        this(
+            dataStartEpoch,
+            dataEndEpoch,
+            anomalyDetector,
+            xContent,
+            client,
+            expirationEpochMs,
+            Clock.systemUTC(),
+            settings,
+            maxEntitiesPerInterval,
+            pageSize
+        );
+    }
+
+    public void start(ActionListener<Page> listener) {
+        try {
+            RangeQueryBuilder rangeQuery = new RangeQueryBuilder(anomalyDetector.getTimeField())
+                .gte(dataStartEpoch)
+                .lt(dataEndEpoch)
+                .format("epoch_millis");
+
+            BoolQueryBuilder internalFilterQuery = new BoolQueryBuilder().filter(anomalyDetector.getFilterQuery()).filter(rangeQuery);
+
+            // multiple categorical fields are supported
+            CompositeAggregationBuilder composite = AggregationBuilders
+                .composite(
+                    AGG_NAME_COMP,
+                    anomalyDetector
+                        .getCategoryField()
+                        .stream()
+                        .map(f -> new TermsValuesSourceBuilder(f).field(f))
+                        .collect(Collectors.toList())
+                )
+                .size(pageSize);
```

Comment:
Is the page size hardcoding somewhere or configurable?

Reply:
It is configurable via a dynamic setting.